docs/dev/migration.md

```diff
@@ -25,9 +25,326 @@ under the License.
+
+    void snapshotState(FunctionSnapshotContext context) throws Exception;
+
+    void initializeState(FunctionInitializationContext context) throws Exception;
+
+The difference here is that although, as in Flink-1.1, the `snapshotState()` is called when a checkpoint 
+is performed, the `initializeState()` which is the counterpart of the `restoreState()`, is called every time 
+the user-defined function is initialized and not only in the case that we are recovering from a failure. 
+Given this, the `initializeState()` is the place where the different types of the state can be initialized, 
+but also where the state recovery logic is included. The code for the `BufferingSink` implementing the 
+`CheckpointedFunction` interface is presented below.
+
+    public class BufferingSink implements SinkFunction<Tuple2<String, Integer>>,
+            CheckpointedFunction, CheckpointedRestoring<ArrayList<Tuple2<String, Integer>>> {
+
+        private final int threshold;
+
+        private transient ListState<Tuple2<String, Integer>> checkpointedState;
+
+        private List<Tuple2<String, Integer>> bufferedElements;
+
+        public BufferingSink(int threshold) {
+            this.threshold = threshold;
+            this.bufferedElements = new ArrayList<>();
+        }
+
+        @Override
+        public void invoke(Tuple2<String, Integer> value) throws Exception {
+            bufferedElements.add(value);
+            if (bufferedElements.size() == threshold) {
+                for (Tuple2<String, Integer> element: bufferedElements) {
+                    // send it to the sink
+                }
+                bufferedElements.clear();
+            }
+        }
+
+        @Override
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
+            checkpointedState.clear();
+            for (Tuple2<String, Integer> element : bufferedElements) {
+                checkpointedState.add(element);
+            }
+        }
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {
+            checkpointedState = context.getOperatorStateStore().
+                getSerializableListState("buffered-elements");
+
+            if (context.isRestored()) {
+                for (Tuple2<String, Integer> element : checkpointedState.get()) {
+                    bufferedElements.add(element);
+                }
+            }
+        }
+
+        @Override
+        public void restoreState(ArrayList<Tuple2<String, Integer>> state) throws Exception {
+            // this is from the CheckpointedRestoring interface.
+            this.bufferedElements.addAll(state);
+        }
+    }
+
+The `initializeState` takes as argument a `FunctionInitializationContext`. This is used to initialize 
+the non-keyed state "container". This is a container of type `ListState` where the non-keyed state objects 
+are going to be stored upon checkpointing:
```

Comment:
This is a container of type `ListState` where non-keyed state objects are stored upon checkpointing: